symbolic/src/Data/Macaw/Symbolic/Memory.hs

```diff
@@ -242,25 +243,68 @@ newGlobalMemoryWith
  -> MC.Memory (MC.ArchAddrWidth arch)
  -- ^ The macaw memory
  -> m (CL.MemImpl sym, MemPtrTable sym (MC.ArchAddrWidth arch))
-newGlobalMemoryWith hooks proxy bak endian mmc mem = do
+newGlobalMemoryWith hooks proxy bak endian mmc mem =
+  newMergedGlobalMemoryWith hooks proxy bak endian mmc [mem]
+
+-- | A version of 'newGlobalMemoryWith' that takes a list of memories and
+-- merges them into a flat addresses space.
+newMergedGlobalMemoryWith
+ :: ( 16 <= MC.ArchAddrWidth arch
+    , MC.MemWidth (MC.ArchAddrWidth arch)
+    , KnownNat (MC.ArchAddrWidth arch)
+    , CB.IsSymBackend sym bak
+    , CL.HasLLVMAnn sym
+    , MonadIO m
+    , sym ~ WEB.ExprBuilder t st fs
+    , ?memOpts :: CL.MemOptions
+    )
+ => GlobalMemoryHooks (MC.ArchAddrWidth arch)
+ -- ^ Hooks customizing the memory setup
+ -> proxy arch
+ -- ^ A proxy to fix the architecture
+ -> bak
+ -- ^ The symbolic backend used to construct terms
+ -> CLD.EndianForm
+ -- ^ The endianness of values in memory
+ -> MemoryModelContents
+ -- ^ A configuration option controlling how mutable memory should be represented (concrete or symbolic)
+ -> [MC.Memory (MC.ArchAddrWidth arch)]
+ -- ^ The macaw memories
+ -> m (CL.MemImpl sym, MemPtrTable sym (MC.ArchAddrWidth arch))
+newMergedGlobalMemoryWith hooks proxy bak endian mmc mems = do
   let sym = CB.backendGetSym bak
-  let ?ptrWidth = MC.memWidth mem
+
+  let ?ptrWidth =
+        case map MC.memWidth mems of
+          [] -> error "Cannot build merged global memory from empty list"
+          x : xs ->
+            -- Check that all memories have the same pointer width
+            if all (testNatEq x) xs
```

Comment:
Don't we already have a proof that the widths are all `ArchAddrWidth arch` (by virtue of the type of the list)? That would make the test unnecessary